Test

simonmarwitz opened the pull request:
I dont know what i am doing. Git is just overkill....

akaszynski commented:
Haha!  I like the PR message.

If things get really bad, just start with a clean slate by deleting the forked repo and cloning from a new forked repo.  It looks like you're adding all the commits when you should just be merging with "master".

Git takes a bit of getting used to, but there's no other (good) way of doing collaborative work, especially online.

akaszynski commented:
If you'd still like to contribute, just let me know what you're trying to accomplish and I can implement it.

simonmarwitz commented:
I found some bug in the code, that i commited in november and tried to commit the fixes for it. 

In the method "nodal_time_history", if we work with reduced outputs, i.e. node components (command CM), the record is only "nitems" long (shorter). The subsequent reordering i implemented was buggy. 

Unfortunately, you adopted the ResultFile class to the new binary format and removed the capabilities for taking into account the node components, so my fix is rather obsolete. I tried to re-add this functionality, but as i don't speak c/cpp/cython I stopped. I spent already so much time on it, while i should actually be working on my research. 

Actually the capability for node components could be added to "nodal_solution" (and other functions) as well, or the record length should actually be checked in "void* read_record" (as far as i remember this i indicated in the first and last 4 bits of a record).

I would be glad if you implemented it, if not, this is also OK, and i will just stick with the code i am using right now. I am very grateful already for this great module anyway. Thank you!

akaszynski commented:
If you post the code here I can take a look at it.

simonmarwitz commented:
Thanks for taking my problem into consideration: I'll try to give you some explanation of what I did previously. I tried to look into the binary_reader.cpp how I could implement it there, but unfortunately, I really do not have any idea. It actually looks like it is implemented somehow.

Consider the following [file.rst.zip](https://github.com/akaszynski/pyansys/files/4274153/file.rst.zip):
It contains 3 timesteps of a transient analysis with a reduced number of nodes: only nodes 20 and 30 are present. The displacements, velocities and accelerations in z direction are all stored in the file. The new rst.py code fails on this file.

As described in fddesc.inc, each record contains 

> 2 integer words added to the front and 1 integer word added to the back. The first word is an integer count that tells how long the record is in 4 byte integer words.  The second word is the high order part of the first word, with the bit 31 set if the record contains integers, clear if the record contains double precision words. The last word is a repeat of the first 4 byte integer word on the front of the record.

My old code (based on pyansys 0.39.4) took care of this in the following way:

```python
f.seek((rpointers[rnum] + ptrSL) * 4)

# integer count that tells how long the record is
reclenl = np.fromfile(f,endian+'i',1)[0]
# read as unsigned integer to to read and reset 31st bit
reclenh = np.fromfile(f, endian+'u4',1)[0]

# bit 31 is set in higher order part, if the record contains integers
is_integer= bool(reclenh & 0b1<<31)
dtype = ['d','i'][is_integer]
#reset 31st bit and represent as int32
reclenh = np.int32(reclenh&~0b1<<31)

if not is_integer:# double count
    nitems = int(two_ints_to_long(reclenl, reclenh)/2)
else:# integer count
    nitems = two_ints_to_long(reclenl, reclenh)

result = np.fromfile(f, endian + dtype, nitems)
```

Later, it is necessary to extract the the set of nodes that was output in order to return them in the correct order. in fdresu.inc: in * NSL

> Note 2: If, upon reading of this record, there is less than nnod*Sumdof items in the
>  record, then only a selected set of nodes were output. Another record follows (integer, less than nnod long) which contains the list of nodes for which DOF solutions are available.  (bit 10 (PDBN) in mask)

which seems to apply to velocities and accelerations, as well.
In my code, this following record was read and reordering, based on the list of reduced nodes, was done. Actually, here i made a mistake, which is fixed in the posted code.
```python
# PDBN should be set if only a subset of nodes was output
# PDBN is set only when solution type: nodal solution/displacement
# PDBN is not set when solution type: acceleration, velocities 
if nitems != Sumdof*nnod:
    # integer count that tells how long the record is
    reclenl = np.fromfile(f,endian+'i',1)[0]
    # read as unsigned integer to read and toogle 31st bit
    reclenh = np.fromfile(f, endian+'u4',1)[0]

    # bit 31 is set in higher order part, if the record contains integers
    is_integer= bool(reclenh & 0b1<<31)
    dtype = ['d','i'][is_integer]
    #reset 31st bit and represent as int32
    reclenh = np.int32(reclenh&~0b1<<31)

    if not is_integer:# double count
        nitems = int(two_ints_to_long(reclenl, reclenh)/2)
    else:# integer count
        nitems = two_ints_to_long(reclenl, reclenh)
    # read record that contains the list of nodes for which DOF solutions are available
    # seems to be in internal ordering
    nodlist = np.fromfile(f,endian+dtype,nitems)#[:,np.newaxis]

    # convert to zero index
    nodlist -= 1

    results[rnum,nodlist,:]=result

    # Reorder based on sorted indexing
    sidx = self.sidx
    results[rnum,:,:] = results[rnum,sidx,:]
else:
    # Reorder based on sorted indexing
    sidx = self.sidx
    results[rnum,:,:]=result.take(sidx,0)
```


With my old code I could run

```python
    res=pyansys.read_binary("file.rst")
    nnum, all_disp = res.nodal_time_history('NSL')
    print(nnum, all_disp[:,:,2])
    nnum, all_vel = res.nodal_time_history('VEL')
    print(nnum, all_vel[:,:,2])
    nnum, all_acc = res.nodal_time_history('ACC')
    print(nnum, all_acc[:,:,2])
```
which should output

```
[10 20 30] [[ 0.          1.          1.        ]
 [ 0.          0.33027225  0.99642433]
 [ 0.         -0.52318946  0.98146012]]
[10 20 30] [[   0.            0.            0.        ]
 [   0.         -446.47411248   -2.38372363]
 [   0.         -122.53026908   -7.59239944]]
[10 20 30] [[      0.               0.               0.        ]
 [      0.         -148824.70416152    -794.57454354]
 [      0.          107981.28113368   -1736.22526957]]
```
